### DISCUSSION TITLE
How to skip all tests from the given test class?

### DISCUSSION THREAD

**Discussion by readonlyden** [xunit/xunit#2580] [Question] [answered]
Hello, we've got a test class for one of our service with ~3000 tests, all tests are marked with `[Fact]` or `[Theory]`.
We want to skip them for a while. Is it possible to do this without manually adding `Skip` to all `[Fact]` or `[Theory]` attributes, and if it is, how to do this correctly, xunit-way?
Thanks!

**Comment by bartelink** [comment, marked as answer]
TL;DR make Test Class private or use a single `const` value for `Skip` at top of file https://stackoverflow.com/a/14845729/11635

**Reply by Xelager** [reply]
>  use a single const value for Skip at top of file https://stackoverflow.com/a/14845729/11635

It will not be possible to apply skip to the entire class. 
`[FactAttribute] [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]`

> make Test Class private

There will be an error from xunit1000: Test classes must be public

**Reply by bartelink** [reply]
>It will not be possible to apply skip to the entire class.
The point is that you have a single const that you can then flip to `null` to re-enable and/or see usages of. I agree you can't do it at class level, and it does involve touching every Theory and/or Fact attribte.
>warning
disable the warning ;)

I don't believe there are any other solutions...

**Reply by readonlyden** [reply]
> > It will not be possible to apply skip to the entire class.
> > The point is that you have a single const that you can then flip to `null` to re-enable and/or see usages of. I agree you can't do it at class level, and it does involve touching every Theory and/or Fact attribte.
> > warning
> > disable the warning ;)
> 
> I don't believe there are any other solutions...

Maybe it's better to create a new attribute, for example, `[SkipAttribute]` and use it in test discoverer? I can try to create a PR for that

**Reply by bartelink** [reply]
V2 is not taking changes of this nature (it's got a massive installed base but energy is going to V3 so critical bugfixes is the bar).

I can't recall whether the 3.x release line is adding (or has added) support for such a facility

Best to do some research as to whether it's been a) done b) discussed and rejected first (to my mind it's a pretty rare case, so I would not take it as a given that the maintainers would necessarily consider such a feature worth adding).

If you're sure it's not been done or previously rejected, I'd create an issue suggesting adding it - that can then serve as the place to discuss doing a PR etc.

**Reply by fvarukhin** [reply]
@bartelink 
I understand your position, you have a lot of more important things to be done. On the other hand, it is not rare case, I believe most of the developers asked this question, it's not a matter of performance or tool quality, it is small feature which makes live of every developer little bit easier. If you google for that question you will find just tons of frustrations and work arounds. As far as I understand @readonlyden he wants to implement it, and together we will do XUnit better for each developer who use it, or plan to use. Just think about it

**Reply by bartelink** [reply]
I don't have a position - I am not the maintainer.

I'm trying to get two things across:
1) the bar for stuff to go into V2 is higher than V3
2) it may already be in V3, or it may have been ruled out

So I'm suggesting that people save themselves and the maintainers time by 
a) making sure that it has not been ruled out
b) (only if a checks out) opening a new issue to discuss the proposed API

Just going off and doing a PR which won't get merged won't be a win for anyone.

In other words, I'm trying to reflect the fact that, in general, the bar for "random feature that <90% of users need" is probably higher in xUnit than in some comparable frameworks. i.e. there's a reason why there are so few attributes in xUnit's core, and it may be worth considering that before making masterplans about "making xunit better" without first 1. checking this has not been traversed 2. (assuming it has not been) writing up a concrete proposal which is easy for the maintainers to give a clear response on 3. (assuming the proposal works out) doing the PR

**Reply by bradwilson** [reply]
I am the maintainer, and I do have positions. 😄 

We do not offer the ability to skip an entire class. We are not currently planning to do this as we believe Skip should be a transient thing. If you are trying to skip 3000 tests temporarily, the best I can offer you is workarounds like commenting out the tests and/or surrounding them with `#if false`/`#endif` during your temporary needs. I don't believe that this behavior is so common (or encouraged) that we would want to add a feature to encourage it. This is a strange thing to do, and it's probably okay that it's painful to do.

**Reply by CodeBlanch** [reply]
@bradwilson Here is use case for consideration...

I have this thing: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Exporter.Geneva

It is an OpenTelemetry component which exports/sends off telemetry about the running application. On Windows it might write to ETW. On Linux it might write to user_events. It supports a bunch of different mechanisms/transports.

I have test classes to cover the different mechanisms. A lot/all of the logic in those classes will only work on a specific platform (Windows or Linux). What would be nice is if I could tell these classes to skip all their tests if not running on the required platform. I know I can do this with custom Fact/Theory attributes but I have to apply those to each test which makes it a bit redundant/ugly/painful. Just sayin' 😄

**Reply by bradwilson** [reply]
You can do this in v3, supplementing `Skip` on your `[Fact]` or `[Theory]` with `SkipWhen` or `SkipUnless`: https://xunit.net/docs/getting-started/v3/whats-new#dynamically-skippable-tests

There is an equivalent code-based way of doing it as well (`Assert.Skip`, `Assert.SkipWhen`, and `Assert.SkipUnless`): https://xunit.net/docs/getting-started/v3/whats-new#dynamic-skipping

**Reply by bradwilson** [reply]
The assertion method is just throwing any exception where the message starts with `$XunitDynamicSkip$` (and the rest of the message is the skip reason), so you can even do that if you don't use our assertion library.

I have used this technique to do an in-one-place environmental skip. In my case, it's to skip trying to run any F# acceptance tests on Mono, since that's not supported by the F# team: https://github.com/xunit/xunit/blob/3e9218416dd0e75ea62388b21c7f24d535daf383/src/xunit.v3.runner.utility.tests/Compilation/FSharpAcceptanceTestAssembly.cs#L24-L25

You could do such a thing via the test class constructor, if all tests in the class need the same skip logic. It can also be done in the constructor of a fixture, if that's a convenient place to do it as well. (And substitute `InitializeAsync` for constructor, if you're doing `IAsyncLifetime`.)
